core/src/test/scala/org/apache/spark/util/collection/OpenHashSetSuite.scala

```diff
@@ -269,4 +269,35 @@ class OpenHashSetSuite extends SparkFunSuite with Matchers {
       assert(pos1 == pos2)
     }
   }
+
+  test("SPARK-45599: 0.0 and -0.0 are equal but not the same") {
```

Comment:
This is a bit tricky and it's better if we can find a reference system that defines this semantic. In Spark, `0.0 == -0.0`, and in GROUP BY, 0.0 and -0.0 are considered to be in the same group and normalized to 0.0.

Reply:
Yes, probably make sense if we just fix this particular issue as `OpenHashSet` is used at many other places.

Reply:
> This is a bit tricky and it's better if we can find a reference system that defines this semantic.

```scala
scala> import java.util.HashSet
import java.util.HashSet

scala> val h = new HashSet[Double]()
val h: java.util.HashSet[Double] = []

scala> h.add(0.0)
val res0: Boolean = true

scala> h.add(-0.0)
val res1: Boolean = true

scala> h.size()
val res2: Int = 2
```

The doc for [HashSet.add][1] states:

> More formally, adds the specified element e to this set if this set contains no element e2 such that Objects.equals(e, e2). If this set already contains the element, the call leaves the set unchanged and returns false.

In other words, `java.util.HashSet` uses `equals` and not `==`, and therefore it considers `0.0` and `-0.0` distinct elements.

So this PR brings `OpenHashSet` more in line with the semantics of `java.util.HashSet`.

[1]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/HashSet.html#add(E)

Reply:
>  In Spark, 0.0 == -0.0, and in GROUP BY, 0.0 and -0.0 are considered to be in the same group and normalized to 0.0.

This PR does not change this behavior. I noticed, however, that we do not have any tests currently to check that -0.0 is normalized and grouped as you describe, so I went ahead and added such a test in 2bfc60548db2c41f4c64b63d40a2652cb22732ab.

Does this address your concern? Or are you suggesting that we should normalize -0.0 to 0.0 across the board?

Reply:
Consider another interesting case where `java.util.HashSet` and `OpenHashSet` differ:

```scala
scala> val h = new HashSet[Double]()
val h: java.util.HashSet[Double] = []

scala> h.add(Double.NaN)
val res9: Boolean = true

scala> h.add(Double.NaN)
val res10: Boolean = false

scala> h.size()
val res11: Int = 1
```

On `master`, `OpenHashSet` does IMO the wrong thing:

```scala
val set = new OpenHashSet[Double]()
set.add(Double.NaN)
set.add(Double.NaN)
set.size  // returns 2
```

This could possibly lead to a bug like the one reported in SPARK-45599 but in reverse, where a new NaN row is added rather than dropped. I will see if I can construct such a scenario as a demonstration. But regardless, I think this behavior is incorrect by itself.

Reply:
Note also that the docstring for `OpenHashSet` seems to imply that it is meant to be a faster but semantically equivalent alternative to `java.util.HashSet`:

https://github.com/apache/spark/blob/0d5c2ce427e06adfebf47bc446ff6646513ae750/core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala#L31-L32

If that's true, then we should perhaps add property based tests to ensure alignment between the two implementations, but I'll leave that as a potential future improvement.

Reply:
Spark's `OpenHashSet` does not have to match `java.util.HashSet`. What matters is the SQL semantic. Can you highlight which functions/operators are using this `OpenHashSet` and what is the impact of this change to the SQL semantic?

Reply:
> Spark's `OpenHashSet` does not have to match `java.util.HashSet`. What matters is the SQL semantic.

Whether or not `OpenHashSet` matches `java.util.HashSet`, I want to emphasize for the record that `OpenHashSet` mishandles 0.0/-0.0 and NaN. Its behavior is simply _incorrect_. [These][test1] [tests][test2] fail on `master` in ways that can only be described as bugs, regardless of whatever SQL semantics we want to preserve.

[This comment][why] explains the root cause. Basically, it is a mistake to combine hash code-based lookups with cooperative equality, at least in the way we are doing it in `OpenHashSet`.

[test1]: https://github.com/apache/spark/pull/45036/files#diff-09400ec633b1f1322c5f7b39dc4e87073b0b0435b60b9cff93388053be5083b6R274-R278
[test2]: https://github.com/apache/spark/pull/45036/files#diff-894198a5fea34e5b7f07d0a4641eb09995315d5de3e0fded3743c15a3c8af405R308-R309
[why]: https://github.com/apache/spark/pull/45036/files#diff-894198a5fea34e5b7f07d0a4641eb09995315d5de3e0fded3743c15a3c8af405R277-R283

But I understand what you are saying. Fixing bugs in `OpenHashSet` doesn't help us if it also breaks users' SQL.

> Can you highlight which functions/operators are using this `OpenHashSet` 

I've updated the PR description with a summary of what uses `OpenHashSet`.

As a side note, I believe that if we accept the change proposed here, we should be able to eliminate `SQLOpenHashSet`. `SQLOpenHashSet` was created specifically to work around the bugs in `OpenHashSet` that we are addressing in this PR. See #33955 and #33993.

> and what is the impact of this change to the SQL semantic?

I've updated the PR description with a diff of what tests pass or fail on `master` vs. this branch. Please take a look and let me know if you think we need any more tests. I know we are touching a sensitive code path and I appreciate the need for caution.